README.md

```diff
@@ -1019,6 +1019,255 @@ Voilà! Since we updated `index.html`, you'll need to restart the dev server. Go
+  target: Dom.element,
+};
+
+type observer;
+[@mel.send] external observe: (observer, Dom.element) => unit = "observe";
+[@mel.send] external unobserve: (observer, Dom.element) => unit = "unobserve";
+[@mel.send] external disconnect: observer => unit = "disconnect";
+
+[@mel.new] external make: (array(entry) => unit) => observer = "IntersectionObserver";
+```
+
+#### 11.2: Update state for pagination
+
+We’ll update our state to track the current page of the feed. We’ll also store
+the previous feed when in the `Loading` state.
+
+```reason
+type step =
+  | Idle /* There is no request going on, and no data to be shown. In this case, we will just show instructions to proceed with the request. */
+  | Loading(option(result(Feed.feed, string))) /* A request is currently taking place to fetch the feed. Stores the previously fetched feed. */
+  | InvalidUsername /* The entered username is not a valid GitHub ID. */
+  | Loaded(result(Feed.feed, string)); /* A request has finished, its result is contained inside the variant value itself. */
+
+type state = {
+  username: string,
+  step,
+  currentPage: int,
+};
+
+let (state, setState) =
+  React.useState(() => {username: "jchavarri", step: Idle, currentPage: 1});
+```
+
+#### 11.3: Add helper functions `mergeFeeds` and `renderFeed`
+
+We are going to be needing these two functions, both can be placed outside the
+`make` function of `App.re`.
+
+The first one takes the old feed and a new one, and proceeds to merge both:
+
+```reason
+let mergeFeeds = (oldFeed: Feed.feed, newFeed: Feed.feed): Feed.feed => {
+  let combinedEntries = Array.concat([oldFeed.entries, newFeed.entries]);
+  {entries: combinedEntries};
+};
+```
+
+The second one is just taking the feed render code and moving it to a separate
+function so we can reuse it:
+
+```reason
+let renderFeed = (feed: Feed.feed) =>
+  <div>
+    <h1> {React.string("GitHub Feed")} </h1>
+    <ul>
+      {switch (feed.entries) {
+       | [||] => React.string("This user feed is empty")
+       | entries =>
+         entries
+         |> Array.map((entry: Feed.entry) =>
+              <li key={entry.id}>
+                <h2> {React.string(entry.title)} </h2>
+                {switch (entry.content) {
+                 | None => React.null
+                 | Some(content) =>
+                   <p dangerouslySetInnerHTML={"__html": content} />
+                 }}
+              </li>
+            )
+         |> React.array
+       }}
+    </ul>
+  </div>;
+```
+
+#### 11.4: Modify `fetchFeed` to support pagination
+
+Update `fetchFeed` to accept a `page` parameter and include it in the API
+request. We’ll use [labeled
+arguments](https://reasonml.github.io/docs/en/function#named-arguments) to avoid
+confusion when calling this function:
+
+```reason
+let fetchFeed = (~username, ~page) => {
+  setState(state =>
+    {
+      ...state,
+      step:
+        switch (state.step) {
+        | Loaded(r)
+        | Loading(Some(r)) => Loading(Some(r))
```

Comment:
```suggestion
        | Loaded(r) => Loading(Some(r))
        | Loading(Some(r)) => Loading(None) // unreachable
```
Since it seems it's not OK to fetch while it's already in the loading state

Reply:
I don't think this is a change we want to apply... Changing the payload of `Loading` from `Some` to `None` means losing whatever feed we had already stored.

The current code is handling the case where `fetchFeed` is inadvertently called when the request is already ongoing, I'm not sure I understand why we want to ditch the previous `feed` state in that case.

Reply:
I think it's still a little confusing, because `fetchFeed` shouldn't be called if `step` is `Loading(_)`:

```reason
          let shouldFetch =
            switch (state.step) {
            | Loading(_) => false
            | Loaded(Ok({entries: [||]})) => false
            | Idle
            | InvalidUsername
            | Loaded(_) => true
            };
          let observer =
            IntersectionObserver.make(entries => {
              let entry = entries[0];
              if (entry.isIntersecting && shouldFetch) {
                fetchFeed(~username, ~page=state.currentPage);
              };
            });
```
But not sure what to change it to, other than raising an error

```suggestion
        | Loaded(r) => Loading(Some(r))
        | Loading(Some(r)) => failwith("fetchFeed shouldn't be called while loading")
```
But that seems to open another can of worms.

Reply:
Yeah, I think if we want to handle all this properly, we should be able to stop the current request when another one happens so there are no race conditions, and properly handle edge cases like that.

For the purposes of this step in the workshop, I think we can go with this approach for now.